### PR TITLE
Support delta brave_installer

### PR DESCRIPTION
**Pull request by simonhong**
In this PR, two new sub options are added to `create_dist` command like below.
`last_chrome_installer` is used to pass folder that contains previous version's uncompressed `chrome.7z` file.
Courgette generate diff file based on previous version's `chrome.7z` file with newly built `chrome.7z`.
```
npm run create_dist Release -- --build_delta_installer --last_chrome_installer="C:/build/installer"
```

This PR should be merged with https://github.com/brave/brave-browser/pull/9667.

Resolves https://github.com/brave/brave-browser/issues/9565

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [x] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: patches/chrome-tools-build-win-create_installer_archive.py.patch [hunk 1]
```diff
@@ -1,33 +1,50 @@
...
++    exe_file = os.path.join(options.build_dir, COURGETTE_EXEC)
+     cmd = '%s -gen "%s" "%s" "%s"' % (exe_file, orig_file, new_file, patch_file)
+   else:
+     exe_file = os.path.join(options.build_dir, BSDIFF_EXEC)
+@@ -179,7 +181,7 @@ def GetPrevVersion(build_dir, temp_dir, last_chrome_installer, output_name):
+                                    output_name + ARCHIVE_SUFFIX)
+   cmd = [lzma_exec,
+          'x',
+-         '-o"%s"' % temp_dir,
++         '-o%s' % temp_dir,
```
Comment (simonhong) [inline patches/chrome-tools-build-win-create_installer_archive.py.patch:35]: W/o this change, `lzma_exec` generates temp_dir under `c:/`.

File: patches/chrome-tools-build-win-create_installer_archive.py.patch [hunk 2]
```diff
@@ -1,33 +1,50 @@
...
  
    if include_snapshotblob == '1':
      CopySectionFilesToStagingDir(config, 'SNAPSHOTBLOB', staging_dir, build_dir,
-@@ -551,6 +552,7 @@ def main(options):
+@@ -155,7 +157,7 @@ def CopySectionFilesToStagingDir(config, section, staging_dir, src_dir,
+ 
+ def GenerateDiffPatch(options, orig_file, new_file, patch_file):
+   if (options.diff_algorithm == "COURGETTE"):
+-    exe_file = os.path.join(options.last_chrome_installer, COURGETTE_EXEC)
++    exe_file = os.path.join(options.build_dir, COURGETTE_EXEC)
```
Comment (simonhong) [inline patches/chrome-tools-build-win-create_installer_archive.py.patch:26]: last_chrome_installer is directory that contains previous version's installer dir.
courgettes exe is generated in ./out.

File: patches/chrome-tools-build-win-create_installer_archive.py.patch (hunk 1 above)
Comment (bsclifton) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch:35]: Is it ever possible for `temp_dir` to have spaces? If so, you might try something like `-o.\"' % temp_dir`

Comment (simonhong) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch]: Yeah, thanks for catching. It would be much safer.
Changed to `-o./"%s" % temp_dir`.

File: patches/chrome-tools-build-win-create_installer_archive.py.patch (hunk 2 above)
Comment (simonhong) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch:26]: Or we can avoid this patch if we copy `courgette.exe` from `out/Release` to folder indicated by `last_chrome_installer` before running `build_delta_installer`. cc @mihaiplesa

Comment (simonhong) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch]: This change is deleted.
`courgette.exe` will be copied from out/Release to `last_chrome_installer` by pipeline.

File: patches/chrome-tools-build-win-create_installer_archive.py.patch [hunk 3]
```diff
@@ -1,33 +1,41 @@
...
  
    if include_snapshotblob == '1':
      CopySectionFilesToStagingDir(config, 'SNAPSHOTBLOB', staging_dir, build_dir,
-@@ -551,6 +552,7 @@ def main(options):
+@@ -179,7 +181,7 @@ def GetPrevVersion(build_dir, temp_dir, last_chrome_installer, output_name):
+                                    output_name + ARCHIVE_SUFFIX)
+   cmd = [lzma_exec,
+          'x',
+-         '-o"%s"' % temp_dir,
++         '-o./"%s"' % temp_dir,
```
Comment (bridiver) [inline patches/chrome-tools-build-win-create_installer_archive.py.patch:26]: why is this change needed?

File: BUILD.gn [hunk 4]
```diff
@@ -59,6 +59,15 @@ group("brave") {
...
     ]
   }
+
+  if (is_win && is_official_build) {
+    # Generate courgette binary by brave target instead of create_dist target.
+    # courgette.exe should be copied to the folder pointed by last_chrome_installer
+    # before running create_dist with build_delta_installer.
+    # When mini_installer target generates diff installer patch, it assumes
+    # that courgette.exe is already in last_chrome_installer folder.
+    deps += [ "//courgette:copy_courgette_binaries" ]
```
Comment (bridiver) [inline BUILD.gn:69]: I don't think this is the right place to put this. https://github.com/brave/brave-core/pull/5484/files#diff-769a945034c01d9bc9d33ecfa819c902R20 should be consolidated into a single dep on `//brave/installer/mini_installer` which has deps for `courgette:copy_courgette_binaries` conditionally on `build_delta_installer` and then on `//brave` instead of `//brave:packed_resources` and `//brave/build/win:copy_exe`

Comment (bridiver) [reply BUILD.gn:69]: also this list is incomplete and can be eliminated by just making the target dependent on `//brave` https://github.com/brave/brave-core/pull/5484/files#diff-769a945034c01d9bc9d33ecfa819c902R19

File: patches/chrome-tools-build-win-create_installer_archive.py.patch (hunk 3 above)
Comment (simonhong) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch:26]: Mentioned here - https://github.com/brave/brave-core/pull/5484#discussion_r422034439

File: patches/chrome-tools-build-win-create_installer_archive.py.patch (hunk 1 above)
Comment (bridiver) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch]: I still don't understand why this is needed, how does it work for chrome?

File: patches/chrome-tools-build-win-create_installer_archive.py.patch (hunk 3 above)
Comment (bridiver) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch:26]: I think the problem here is what we're passing in for `last_chrome_installer` because we shouldn't have to change the script itself. We should be passing in a a rebased value here https://github.com/brave/brave-core/pull/5484/files#diff-769a945034c01d9bc9d33ecfa819c902R18

Comment (bridiver) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch:26]: sorry, I guess this is `staging_dir`, but in any case this should be running in the root build dir so there's no reason for it to end up in the root of the filesystem

File: patches/chrome-tools-build-win-create_installer_archive.py.patch (hunk 1 above)
Comment (simonhong) [reply patches/chrome-tools-build-win-create_installer_archive.py.patch:26]: After discussing with @bridiver , This is reverted to `'-o%s' % temp_dir,` because relative path in out doesn't have space.
